### ISSUE TITLE
Can't alt click to move cursor

### ISSUE THREAD

**Issue by pdufour** [zeit/hyper#316]
Terminal app on mac has a nice feature where you can `alt + click` to move the cursor to a position. This does not work in Hyperterm.

**Comment by timneutkens** [comment]
@pdufour by just looking through hyperterm issues my mind gets blown by the crazy time saving things I didn't know about iTerm and mac terminal. Thanks for sharing :wink: +1

**Comment by yaycmyk** [comment]
This is a must-have feature.

**Comment by josiahwiebe** [comment]
Here's a gif for reference.

![2016-10-27 08_26_56](https://cloud.githubusercontent.com/assets/265638/19768966/2b48db3a-9c1f-11e6-9c6e-9a1e35ceabb3.gif)

**Comment by ppot** [comment]
The way hterm interpret it with span doesn't allow us the ability to do this now. From what I tried. I can set the cursor position but then it doesn't comply to text after it.

**Comment by matheuss** [comment]
Copying what I wrote on #1466: 
> that's a thing I'd be very interested to see on Hyper. I'd be very interested on implementing it too, but unfortunately I'm incredibly busy lately 😔

**Comment by adamscybot** [comment]
Is this possible with a plugin? @ppot Can you give more detail on why it isn't doable?

Would like to have a go at this unless the API totally prevents it.

**Comment by ppot** [comment]
@adamscybot From what I tried, you can update the cursor position, but it didn't sync. You can try!

**Comment by adamscybot** [comment]
Ah as in, the server didn't actually recognize the change? I guess you'd have to send the right control characters over the ssh to move the cursor, i.e. the equivalent of pressing the right arrow key however many keys you need to move.

**Comment by adamscybot** [comment]
Ok, I have a working solution that is very very hacky. As mentioned, the lack of spans to detect click event makes this very tricky. I think proper support would need to be integrated into hyper itself. Or an improved plugin API.

Regardless, I know it will help some people so I will finish it off this week and publish (hopefully tomorrow evening, GMT).

**Comment by adamscybot** [comment]
Ok guys here we go: https://github.com/adamscybot/hyper-alt-click

It's a giant hack, but I just *needed* this feature. Very experimental at this stage, hopefully room for improvement. Especially support in Vim/Nano etc.

Completely untested on everything but my mac...

**Comment by chabou** [comment]
@adamscybot Which API improvements do you need to achieve this with a less-hacky plugin?

**Comment by adamscybot** [comment]
@chabou The main issue is that each line in the terminal exists as a text node in the DOM. This means you can not attach a click handler to an individual letter. I think there needs to be a way to pass down your own element for rendering a line in the terminal. Or alternatively, the `<span>`'s are added and there is a new cursor position prop somewhere.

Im hacking around this problem by:

- Detecting a click event on the whole line
- Duplicating the text, but every character wrapped in a span.
- Overlay this duplication directly over the top.
- Add handlers to these spans.
- Refire the click event to see which letter was clicked on.
- Play around with offsets (also hacky) to know how many times to send the left or right control character.

I believe the base issue though actually exists in Google Hterm though, which seems to be responsible for the handling of the actual terminal lines. Not sure if it can easily be hacked in or if hterm would need to be replaced.

**Comment by chabou** [comment]
Adding a span to every char could have a huge impact on performances.
Your hack has a great advantage: it takes place only after a `alt + click` event 😛 

Hyper has a lot of hterm monkey patches. You can look deeper into hterm, try to add span for each chars and measure performance changes.

Hyper has already a lot of performances issues. Adding DOM complexity seems harmful.

**Comment by adamscybot** [comment]
Yeh, I can see how that could be an issue. Perhaps one route would be to try and formalise my method in some way? If there was a way to attach a click handler on a terminal line and then to modify how lines are rendered (pass down a custom component?) on a per-line basis.

**Comment by chabou** [comment]
Hterm is itself outside react ecosystem (encapsulated in an iFrame).
And imho, it is a bad idea that Hyper hacks and exposes some hterm's low-level mechanisms. Plugin should do it by itself using onTerminal() hook to access hterm instance

**Comment by watzon** [comment]
I was going to make a new issue, but this is related. In linux and OSX you can click to move the cursor in Vim and other apps. I know it's not a best practice for vim, but sometimes it's nice when I'm in a "just f*ck it" mood and don't want to deal with keyboard shortcuts.

**Comment by chabou** [comment]
You're right.
In Terminal app, I have to `alt+click` to move on command line. But it works fine without `alt` in `vim` (with my config). It doesn't work in Hyper.

**Comment by chabou** [comment]
The good news is: it works great in `vim` in our `xterm` branch (replacing `hterm` by `xterm.js`)

**Comment by adamscybot** [comment]
@chabou Oooo xterm. Does this bring anything to the table in terms of per-character mouse events? Is alt+click already supported out of the box?

**Comment by chabou** [comment]
@adamscybot col/row notion is reaaaally better handled in `xterm.js`. You can take a look to selection helpers https://github.com/sourcelair/xterm.js/blob/master/src/utils/Mouse.ts. 
`Alt+click` is not supported yet, but it is not like `hterm`: `xterm.js` is really open-source and PRs are welcome. @Tyriar will certainly be ok to have a PR to implement this `alt+click` directly in `xterm.js` because it will be great to have it in VSCode too.

**Comment by adamscybot** [comment]
Ah this is exactly what was needed. I'm guessing I could just grab the xterm instance in a hyper plugin and plug away at it. I'll take a look once its in master.

**Comment by Tyriar** [comment]
@adamscybot yes PRs welcome in xterm.js. If you want to contribute in this area please fork the [v3 branch](https://github.com/sourcelair/xterm.js/tree/v3) and make changes there to avoid merge conflict headaches. Created https://github.com/sourcelair/xterm.js/issues/890 to discuss implementation 😃 

Also working against the demo https://github.com/sourcelair/xterm.js#demo will be easier than trying to pull the right version in to hyper.

**Comment by CodeTheory** [comment]
Update: There's a PR in xterm for this feature https://github.com/xtermjs/xterm.js/pull/896

**Comment by Tyriar** [comment]
And we merged it today, coming in the next version 😃
